### PR TITLE
Simplify the DB schema

### PR DESCRIPTION
**Pull request by simlmx**
Remove the LatestForecastValues table
* Remove the DatetimeInterval table
* Use CamelCase sqlalchemy types when possible (everything except UUID)
* Add a foreign key from ForecastValues to Sites
* Add indexes on some foreign keys and `start_utc` columns
* Add an index on client_site_id and client_site_name
* Remove most `write` functions: clients of this library are responsible
  to write their own helper functions (only keeping the "generation" one
  because the change would be too involved).
* Rename power_kw to {generation,forecast}_power_kw
* Remove default UUIDs for non-primary keys

Addresses part of #48 

@peterdudfield and @devsjc, feel free to comment/challenge.

### CODE REVIEW

File: sdk/python/pvsite_datamodel/sqlmodels.py [hunk 1]
```diff
@@ -20,7 +18,7 @@
 class CreatedMixin:
     """Mixin to add created datetime to model."""
 
-    created_utc = Column(DateTime(timezone=True), default=lambda: datetime.utcnow())
```
Comment (simlmx) [inline sdk/python/pvsite_datamodel/sqlmodels.py:23]: I removed the timezone information for simplicity and consistency with the other DateTime fields.

File: sdk/python/pvsite_datamodel/sqlmodels.py [hunk 2]
```diff
@@ -38,27 +36,34 @@ class SiteSQL(Base, CreatedMixin):
...
     dno = sa.Column(sa.String(255))
     gsp = sa.Column(sa.String(255))
-    orientation = sa.Column(REAL)
-    tilt = sa.Column(REAL)
-    latitude = sa.Column(FLOAT, nullable=False)
-    longitude = sa.Column(FLOAT, nullable=False)
-    capacity_kw = sa.Column(REAL, nullable=False)
-    updated_utc = sa.Column(TIMESTAMP, nullable=False, default=lambda: datetime.utcnow())
+
+    # For metadata `NULL` means "we don't know".
```
Comment (simlmx) [inline sdk/python/pvsite_datamodel/sqlmodels.py:51]: Having possible `NULL` values for metadata seemed like a requirement.

File: sdk/python/pvsite_datamodel/sqlmodels.py [hunk 3]
```diff
@@ -128,60 +138,33 @@ class ForecastValueSQL(Base, CreatedMixin):
...
-        nullable=False,
-        default=uuid.uuid4,
+    forecast_value_uuid = sa.Column(
+        UUID(as_uuid=True), default=uuid.uuid4, primary_key=True
     )
-    forecast_generation_kw = sa.Column(REAL, nullable=False)
+
+    start_utc = sa.Column(sa.DateTime, nullable=False, index=True)
+    end_utc = sa.Column(sa.DateTime, nullable=False)
+    power_kw = sa.Column(sa.Float, nullable=False)
```
Comment (simlmx) [inline sdk/python/pvsite_datamodel/sqlmodels.py:147]: I found it very confusing that the predicted power and the historical power had different names, but I'm happy to revert the change if there was a reason for the different name.

Comment (peterdudfield) [reply sdk/python/pvsite_datamodel/sqlmodels.py:147]: I thought it was good having slightly different names, so people dont get confused between a forecast and actual

Comment (simlmx) [reply sdk/python/pvsite_datamodel/sqlmodels.py:147]: Yeah but since the table name is never going to be far, perhaps it's not that confusing? But what was even more confusing is that it's `Forecasts.forecast_generation_kw` but not `Generation.generation_kw`; at least it should be `Forecasts.forecast_power_kw` or something.

Comment (peterdudfield) [reply sdk/python/pvsite_datamodel/sqlmodels.py:147]: go for
- `Forecasts.forecast_power_kw`
- `Generation.generation_power_kw`
?

Comment (devsjc) [reply sdk/python/pvsite_datamodel/sqlmodels.py:147]: There's no right answer to this, it's the same as the argument about `table_name.table_uuid` vs `table_name.uuid` - we just have to pick one! Peter's suggestion above seems fine to me.

File: sdk/python/pvsite_datamodel/sqlmodels.py [hunk 4]
```diff
@@ -38,27 +36,33 @@ class SiteSQL(Base, CreatedMixin):
...
+
     region = sa.Column(sa.String(255))
     dno = sa.Column(sa.String(255))
     gsp = sa.Column(sa.String(255))
-    orientation = sa.Column(REAL)
-    tilt = sa.Column(REAL)
-    latitude = sa.Column(FLOAT, nullable=False)
-    longitude = sa.Column(FLOAT, nullable=False)
-    capacity_kw = sa.Column(REAL, nullable=False)
-    updated_utc = sa.Column(TIMESTAMP, nullable=False, default=lambda: datetime.utcnow())
```
Comment (simlmx) [inline sdk/python/pvsite_datamodel/sqlmodels.py:53]: Note that I removed `updated_utc` for now. It wasn't used properly anyway. If we want it we should put it back with a proper postgreql trigger so that it is automatically updated.

File: sdk/python/pvsite_datamodel/read/generation.py [hunk 5]
```diff
@@ -68,19 +72,24 @@ def get_pv_generation_by_sites(
     # start main query
     query = session.query(GenerationSQL)
     query = query.join(SiteSQL)
-    query = query.join(DatetimeIntervalSQL)
 
     # Filter by time interval
-    query = filter_query_by_datetime_interval(query=query, start_utc=start_utc, end_utc=end_utc)
+    if start_utc is not None:
```
Comment (peterdudfield) [inline sdk/python/pvsite_datamodel/read/generation.py:77]: could refactor this, but I actually quite like the readabiliy of having this all here

File: sdk/python/pvsite_datamodel/read/latest_forecast_values.py [hunk 6]
```diff
@@ -2,115 +2,46 @@
...
-    """Get the forecast values by input sites, get the lastest value
+    """Get the forecast values by input sites, get the lastest value.
 
     This reads the ForecastValueSQL table
 
     :param session: The sqlalchemy database session
     :param site_uuids: list of site_uuids for which to fetch latest forecast values
     :param start_utc: filters on forecast values target_time >= start_utc
-    :param model, the database table to use, could be 'LatestForecastValueSQL' or ForecastValueSQL
-    :return: dict containing {site_uuid1: List[LatestForecastValueSQL], site_uuid2: ...}
```
Comment (peterdudfield) [inline sdk/python/pvsite_datamodel/read/latest_forecast_values.py:79]: perhaps worth keeping in this return description

Comment (peterdudfield) [reply sdk/python/pvsite_datamodel/read/latest_forecast_values.py:79]: although it is in the typing, so yea, get rid of it

File: sdk/python/pvsite_datamodel/sqlmodels.py (hunk 2 above)
Comment (peterdudfield) [reply sdk/python/pvsite_datamodel/sqlmodels.py:50]: yea, lets get this through, and we can always come back to it. There was a request for `capacity_kw` o be null

File: sdk/python/pvsite_datamodel/sqlmodels.py [hunk 7]
```diff
@@ -73,20 +77,20 @@ class GenerationSQL(Base, CreatedMixin):
...
-        default=uuid.uuid4,
-    )
-    datetime_interval: DatetimeIntervalSQL = relationship(
-        "DatetimeIntervalSQL", back_populates="generation"
+        index=True,
     )
+    generation_power_kw = sa.Column(sa.Float, nullable=False)
+
+    start_utc = sa.Column(sa.DateTime, nullable=False, index=True)
+    end_utc = sa.Column(sa.DateTime)
```
Comment (peterdudfield) [inline sdk/python/pvsite_datamodel/sqlmodels.py:92]: could index on this too? But leave it for the moment

File: sdk/python/pvsite_datamodel/write/generation.py [hunk 8]
```diff
@@ -1,97 +1,73 @@
...
 
-from pvsite_datamodel.read.site import get_site_by_uuid
-from pvsite_datamodel.sqlmodels import GenerationSQL
+from pvsite_datamodel.sqlmodels import Base, GenerationSQL
 
-# Defines the length of time over which a forecast is valid
-from pvsite_datamodel.write.datetime_intervals import get_or_else_create_datetime_interval
-from pvsite_datamodel.write.upsert import upsert
-from pvsite_datamodel.write.utils import UUIDV4, WrittenRow
+_log = logging.getLogger(__name__)
```
Comment (peterdudfield) [inline sdk/python/pvsite_datamodel/write/generation.py:14]: thanks for keeping these in

File: sdk/python/tests/test_connection.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    SiteSQL,
+)
+
+class TestDatabaseConnection:
+    """Tests for the DatabaseConnection class."""
+
+    def test_connection(self, engine, sites):
+        dbcon = DatabaseConnection(engine.url, echo=False)
+        with dbcon.get_session() as session:
+            session.query(SiteSQL).first()
```
Comment (peterdudfield) [inline sdk/python/tests/test_connection.py:13]: thanks for adding this

**Review by peterdudfield — APPROVED** [review]
Looks good, @devsjc perhaps you have a look and then we can merge it in

File: sdk/python/pvsite_datamodel/read/generation.py [hunk 10]
```diff
@@ -68,19 +72,24 @@ def get_pv_generation_by_sites(
...
+        query = query.filter(
+            GenerationSQL.end_utc < end_utc,
+        )
 
     if site_uuids is not None:
         query = query.filter(SiteSQL.site_uuid.in_(site_uuids))
 
-    # Order by 'created_utc' desc
-    query = query.order_by(SiteSQL.site_uuid, DatetimeIntervalSQL.start_utc)
+    query = query.order_by(SiteSQL.site_uuid, GenerationSQL.start_utc)
```
Comment (devsjc) [inline sdk/python/pvsite_datamodel/read/generation.py:90]: Is start_utc in generationSQL indexed?

File: sdk/python/pvsite_datamodel/read/latest_forecast_values.py [hunk 11]
```diff
@@ -2,115 +2,46 @@
...
+from pvsite_datamodel.sqlmodels import ForecastValueSQL
 
 
 def get_forecast_values_by_site_latest(
     session: Session,
     site_uuids: List[uuid.UUID],
     start_utc: Optional[dt.datetime] = None,
 ) -> Dict[uuid.UUID, List[ForecastValueSQL]]:
-    """Get the forecast values by input sites, get the lastest value
+    """Get the forecast values by input sites, get the lastest value.
```
Comment (devsjc) [inline sdk/python/pvsite_datamodel/read/latest_forecast_values.py:17]: Seems to me this function doesn't get the latest value as the name and description state, but rather all values in creation order for the sites

File: sdk/python/pvsite_datamodel/sqlmodels.py [hunk 12]
```diff
@@ -73,20 +77,20 @@ class GenerationSQL(Base, CreatedMixin):
...
         nullable=False,
-        default=uuid.uuid4,
-    )
-    datetime_interval: DatetimeIntervalSQL = relationship(
-        "DatetimeIntervalSQL", back_populates="generation"
+        index=True,
     )
+    generation_power_kw = sa.Column(sa.Float, nullable=False)
+
+    start_utc = sa.Column(sa.DateTime, nullable=False, index=True)
```
Comment (peterdudfield) [inline sdk/python/pvsite_datamodel/sqlmodels.py:91]: lovely index here @devsjc

File: sdk/python/pvsite_datamodel/write/generation.py [hunk 13]
```diff
@@ -1,97 +1,73 @@
...
+    if not update_dict:
+        raise ValueError("insert_or_update resulted in an empty update_dict")
+
+    stmt = stmt.on_conflict_do_update(
+        index_elements=primary_key_names, set_=update_dict
     )
+    session.execute(stmt, rows)
 
 
 def insert_generation_values(
```
Comment (devsjc) [inline sdk/python/pvsite_datamodel/write/generation.py:38]: How come we've lost the schema for the dataframe here?

File: sdk/python/pvsite_datamodel/read/latest_forecast_values.py (hunk 11 above)
Comment (peterdudfield) [reply sdk/python/pvsite_datamodel/read/latest_forecast_values.py:17]: doesnt the `distinct` and `order by` mean it gets the latest values?

File: sdk/python/pvsite_datamodel/write/generation.py (hunk 13 above)
Comment (peterdudfield) [reply sdk/python/pvsite_datamodel/write/generation.py:38]: Yea, I think simon wanted to keep it simon, I personally like the schema checker, so maybe it could be 2:1 on this. @simlmx what do you think?

Pros and cons of keeping schema in
**Pros**
- good validation,
- good validation should lead to clear error messages

**Cons**:
- more code, slight more complicated
- another dependencies in the repo

File: sdk/python/pvsite_datamodel/read/latest_forecast_values.py (hunk 11 above)
Comment (devsjc) [reply sdk/python/pvsite_datamodel/read/latest_forecast_values.py:17]: That just stops duplicate values for the same time in the same site I thought? What I mean to say is the description seems to suggest that it is only going to pull a single value, when in fact it will be returning all the latest values for the site for each start time greater than the input.
Not that it really matters I guess, as long as we know how it works!

File: sdk/python/pvsite_datamodel/read/latest_forecast_values.py (hunk 6 above)
Comment (simlmx) [reply sdk/python/pvsite_datamodel/read/latest_forecast_values.py:79]: Yeah that's why I removed it: it was the same as the typing but had already got out of sync with it.

File: sdk/python/pvsite_datamodel/sqlmodels.py (hunk 7 above)
Comment (simlmx) [reply sdk/python/pvsite_datamodel/sqlmodels.py:92]: Yeah I would wait and see if we do actually need to search on the `end_utc` field. If so, and if we also always filter on `start_utc`, maybe we should consider a multi-column index on (start_utc, end_utc); but I'm not absolutely sure this is actually a right use-cse for this.

File: sdk/python/pvsite_datamodel/write/generation.py (hunk 13 above)
Comment (simlmx) [reply sdk/python/pvsite_datamodel/write/generation.py:38]: First I don't think this code ("write" helper functions) should live in this repo at all: it should be closer to the code using it. I kept it in here for now because it was a tricky refactor. This is for another discussion but I think that is we want different services to interact with the DB in a well defined uniform way, they should probably do it through the API, not through a common helper function. And if two services really need to both interact with the database in a similar fashion, they should probably implement their own helper functions, even if it means having copy-pasted functions at first: those could very well diverge in the future.

We shouldn't need the same kind of validation that we use when receiving data from an external source (like a client sending data to an API) here, for the simple reason that we control the code that creates the data. **We don't need to handle the errors from bad data** (unlike for client-facing APIs), **we simply want to crash because if the data is wrong, it means our own code is broken**. This already happens without any validation: if the type of the data or the column names are wrong, the `INSERT/UPDATE` will fail.

I would also argue that errors like `IndexError: 'column X' not in dataframe` or `DBDataError: column X is expecting a Datetime, Float provided` will be clear enough in this context.

I no one is convinced by my arguments I'm happy to bring the pydantic validation back 😛

File: sdk/python/tests/test_connection.py (hunk 9 above)
Comment (simlmx) [reply sdk/python/tests/test_connection.py:13]: Yeah I actually removed it by accident: it was in the `test_write.py` file.

File: sdk/python/pvsite_datamodel/read/latest_forecast_values.py (hunk 11 above)
Comment (simlmx) [reply sdk/python/pvsite_datamodel/read/latest_forecast_values.py:17]: This is exactly for this sort of things that I think those functions should live in the repos where they are called!

File: sdk/python/pvsite_datamodel/write/generation.py (hunk 13 above)
Comment (peterdudfield) [reply sdk/python/pvsite_datamodel/write/generation.py:38]: Perhaps just put int some asserts for the moment, and we can come back to it another time

Comment (simlmx) [reply sdk/python/pvsite_datamodel/write/generation.py:38]: Yes actually if some *development* errors are easy to make and would not be caught by some other SqlalchemyError/IndexError/etc., then some `assert`s for sanity checks would make a lot of sense.

I think here we are covered by "normal" errors though, or am I forgetting some possible error types?
